# RNN

In [1]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix
from sklearn.preprocessing import StandardScaler
from tensorflow.keras.models import Sequential, Model
from tensorflow.keras.layers import Embedding, LSTM, Dense, Dropout, Input, Concatenate, Reshape, BatchNormalization, Bidirectional
from tensorflow.keras.preprocessing.text import Tokenizer
from tensorflow.keras.preprocessing.sequence import pad_sequences
from tensorflow.keras.regularizers import l2
from keras.callbacks import EarlyStopping, ReduceLROnPlateau
from sentence_transformers import SentenceTransformer

c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\sentence_transformers\cross_encoder\CrossEncoder.py:13: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import tqdm, trange


In [2]:
with open("C:/Yee Ann/NUS/DSA4266/TruthSeeker2023/Features_For_Traditional_ML_Techniques.csv") as file:
    df = pd.read_csv(file)

In [3]:
#remove columns with s single unique value
df = df.drop(columns = ['Unnamed: 0'])
df = df.loc[:, df.nunique() > 1]
df = df.drop_duplicates()

# Convert boolean columns to 0 and 1
df['majority_target'] = df['majority_target'].astype(int)

Basic Tokenizer

In [4]:
# Assuming df is your DataFrame
# Combine the text fields (statement, tweet) into one
df['text_combined'] = df['statement'] + ' ' + df['tweet']

# Tokenize the combined text field
tokenizer = Tokenizer(num_words=5000, oov_token="<OOV>")
tokenizer.fit_on_texts(df['text_combined'])

# Convert text to sequences
X_text = tokenizer.texts_to_sequences(df['text_combined'])
word_index = tokenizer.word_index

In [5]:
# Pad sequences to ensure they are the same length
max_length = 100  # Set max length of sequences
X_text = pad_sequences(X_text, maxlen=max_length, padding='post', truncating='post')

# Separate the numerical features
numerical_features = df.select_dtypes(['int64', 'int32', 'float64']).columns.tolist()
numerical_features.remove('majority_target')
X_num = df[numerical_features].values

# Scale the numerical features
scaler = StandardScaler()
X_num_scaled = scaler.fit_transform(X_num)

In [6]:
# Combine text sequences and numerical features into one dataset
X_combined = np.hstack((X_text, X_num_scaled))

# Define the target variable
y = df['majority_target'].values

# Split the dataset into training and test sets
X_train, X_test, y_train, y_test = train_test_split(X_combined, y, test_size=0.3, random_state=42)

# Build the RNN model
model = Sequential()
model.add(Embedding(input_dim=len(word_index) + 1, output_dim=128, input_length=max_length))  # Embedding layer for text
model.add(LSTM(128, return_sequences=True))  # LSTM layer
model.add(Dropout(0.5))  # Dropout for regularization
model.add(LSTM(64, return_sequences=False))  # Another LSTM layer
model.add(Dropout(0.5))  # More regularization
model.add(Dense(32, activation='relu'))  # Dense layer for additional complexity
model.add(Dropout(0.5))  # Additional Dropout
model.add(Dense(1, activation='sigmoid'))  # Output layer for binary classification

c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\layers\core\embedding.py:90: UserWarning: Argument `input_length` is deprecated. Just remove it.
  warnings.warn(


In [7]:
# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

# Train the model
history = model.fit(X_train, y_train, epochs=10, batch_size=32, validation_split=0.2)

# Evaluate the model
loss, accuracy = model.evaluate(X_test, y_test)
print(f'Test Accuracy: {accuracy}')

# Predict on test data
y_pred = model.predict(X_test)

Epoch 1/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 577s 243ms/step - accuracy: 0.5093 - loss: 0.6939 - val_accuracy: 0.5337 - val_loss: 0.6891
Epoch 2/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 604s 257ms/step - accuracy: 0.5325 - loss: 0.6891 - val_accuracy: 0.5274 - val_loss: 0.6905
Epoch 3/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 605s 258ms/step - accuracy: 0.5404 - loss: 0.6871 - val_accuracy: 0.9270 - val_loss: 0.2698
Epoch 4/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 628s 267ms/step - accuracy: 0.9482 - loss: 0.2253 - val_accuracy: 0.9532 - val_loss: 0.1889
Epoch 5/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 1940s 826ms/step - accuracy: 0.9484 - loss: 0.2227 - val_accuracy: 0.9539 - val_loss: 0.1898
Epoch 6/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 587s 250ms/step - accuracy: 0.9563 - loss: 0.1994 - val_accuracy: 0.9557 - val_loss: 0.1829
Epoch 7/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 562s 239ms/step - accuracy: 0.9557 - loss: 0.2004 - val_accuracy: 0.9557 - val_loss: 0.1848
Epoch 8/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 692s 294ms/step - a

In [9]:
y_pred_use = (y_pred > 0.5).astype(int)
# Evaluate the model using accuracy
accuracy = accuracy_score(y_test, y_pred_use)
print(f"RNN Accuracy: {accuracy:.4f}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred_use, target_names=['Fake', 'Real']))

# Confusion Matrix to understand predictions
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_use))

RNN Accuracy: 0.9559

Classification Report:
              precision    recall  f1-score   support

        Fake       0.95      0.95      0.95     19442
        Real       0.96      0.96      0.96     20818

    accuracy                           0.96     40260
   macro avg       0.96      0.96      0.96     40260
weighted avg       0.96      0.96      0.96     40260


Confusion Matrix:
[[18555   887]
 [  890 19928]]


Sentence Embeddings

In [10]:
# Load pre-trained sentence transformer model
model_sbert = SentenceTransformer('all-MiniLM-L6-v2')

# Generate sentence embeddings for the 'statement' column
X_statement_embeddings = model_sbert.encode(df['statement'].tolist(), convert_to_numpy=True)

# Generate sentence embeddings for the 'tweet' column
X_tweet_embeddings = model_sbert.encode(df['tweet'].tolist(), convert_to_numpy=True)

numerical_features = df.select_dtypes(['int64', 'int32', 'float64']).columns.tolist()
numerical_features.remove('majority_target')
X_num = df[numerical_features].values

# Scale the numerical features
scaler = StandardScaler()
X_num_scaled = scaler.fit_transform(X_num)

modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\huggingface_hub\file_download.py:147: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\laiye\.cache\huggingface\hub\models--sentence-transformers--all-MiniLM-L6-v2. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [11]:
# Define inputs for statement, tweet, and numerical features
input_statement = Input(shape=(X_statement_embeddings.shape[1],))
input_tweet = Input(shape=(X_tweet_embeddings.shape[1],))
input_numerical = Input(shape=(X_num_scaled.shape[1],))

# Reshape to 3D for LSTM (adding a dimension for timesteps)
reshaped_statement = Reshape((1, X_statement_embeddings.shape[1]))(input_statement)  # Shape: (batch_size, 1, 384)
reshaped_tweet = Reshape((1, X_tweet_embeddings.shape[1]))(input_tweet)  # Shape: (batch_size, 1, 384)

# LSTM for statement embeddings
lstm_statement = LSTM(128, return_sequences=False)(reshaped_statement)

# LSTM for tweet embeddings
lstm_tweet = LSTM(128, return_sequences=False)(reshaped_tweet)

# Combine LSTM outputs and numerical features
combined = Concatenate()([lstm_statement, lstm_tweet, input_numerical])

In [12]:
# Add dense layers
# x = Dense(64, activation='relu', kernel_regularizer=l2(0.01))(combined)
x = Dense(64, activation='relu')(combined)
x = Dropout(0.5)(x)
x = Dense(32, activation='relu')(x)
x = Dropout(0.5)(x)
# x = BatchNormalization()(x)
output = Dense(1, activation='sigmoid')(x)

# Define the model
model = Model(inputs=[input_statement, input_tweet, input_numerical], outputs=output)

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [13]:
# Split data into training and test sets
X_train_statement, X_test_statement, X_train_tweet, X_test_tweet, X_train_num, X_test_num, y_train, y_test = train_test_split(
    X_statement_embeddings, X_tweet_embeddings, X_num_scaled, df['majority_target'], test_size=0.3, random_state=42)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=2)
early_stopping = EarlyStopping(monitor='val_loss', patience=3)

# Train the model
history = model.fit([X_train_statement, X_train_tweet, X_train_num], y_train, epochs=10, batch_size=32, validation_split=0.2, callbacks=[early_stopping, reduce_lr])

# Evaluate the model
loss, accuracy = model.evaluate([X_test_statement, X_test_tweet, X_test_num], y_test)
print(f'Test Accuracy: {accuracy}')

# Predict on test data
y_pred = model.predict([X_test_statement, X_test_tweet, X_test_num])

Epoch 1/10


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_9', 'keras_tensor_10', 'keras_tensor_11']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


2349/2349 ━━━━━━━━━━━━━━━━━━━━ 12s 4ms/step - accuracy: 0.8859 - loss: 0.3188 - val_accuracy: 0.9558 - val_loss: 0.1849 - learning_rate: 0.0010
Epoch 2/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9564 - loss: 0.2034 - val_accuracy: 0.9559 - val_loss: 0.1839 - learning_rate: 0.0010
Epoch 3/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9577 - loss: 0.1965 - val_accuracy: 0.9558 - val_loss: 0.1841 - learning_rate: 0.0010
Epoch 4/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9584 - loss: 0.1927 - val_accuracy: 0.9557 - val_loss: 0.1841 - learning_rate: 0.0010
Epoch 5/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9570 - loss: 0.1926 - val_accuracy: 0.9558 - val_loss: 0.1833 - learning_rate: 2.0000e-04
Epoch 6/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9574 - loss: 0.1884 - val_accuracy: 0.9558 - val_loss: 0.1840 - learning_rate: 2.0000e-04
Epoch 7/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 9s 4ms/step - accuracy: 0.9568 - los

In [14]:
y_pred_use = (y_pred > 0.5).astype(int)
# Evaluate the model using accuracy
accuracy = accuracy_score(y_test, y_pred_use)
print(f"RNN Accuracy: {accuracy:.4f}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred_use, target_names=['Fake', 'Real']))

# Confusion Matrix to understand predictions
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_use))

RNN Accuracy: 0.9576

Classification Report:
              precision    recall  f1-score   support

        Fake       0.95      0.96      0.96     19442
        Real       0.96      0.96      0.96     20818

    accuracy                           0.96     40260
   macro avg       0.96      0.96      0.96     40260
weighted avg       0.96      0.96      0.96     40260


Confusion Matrix:
[[18618   824]
 [  883 19935]]


Bidirection RNN

In [15]:
# Define inputs for statement, tweet, and numerical features
input_statement = Input(shape=(X_statement_embeddings.shape[1],))
input_tweet = Input(shape=(X_tweet_embeddings.shape[1],))
input_numerical = Input(shape=(X_num_scaled.shape[1],))

# Reshape to 3D for LSTM (adding a dimension for timesteps)
reshaped_statement = Reshape((1, X_statement_embeddings.shape[1]))(input_statement)  # Shape: (batch_size, 1, 384)
reshaped_tweet = Reshape((1, X_tweet_embeddings.shape[1]))(input_tweet)  # Shape: (batch_size, 1, 384)

# Bidirectional LSTM for statement embeddings
bidirectional_lstm_statement = Bidirectional(LSTM(128, return_sequences=False))(reshaped_statement)

# Bidirectional LSTM for tweet embeddings
bidirectional_lstm_tweet = Bidirectional(LSTM(128, return_sequences=False))(reshaped_tweet)

# Combine LSTM outputs and numerical features
combined = Concatenate()([bidirectional_lstm_statement, bidirectional_lstm_tweet, input_numerical])

In [16]:
# Add dense layers
# x = Dense(64, activation='relu', kernel_regularizer=l2(0.01))(combined)
x = Dense(64, activation='relu')(combined)
x = Dropout(0.5)(x)
x = Dense(32, activation='relu')(x)
x = Dropout(0.5)(x)
# x = BatchNormalization()(x)
output = Dense(1, activation='sigmoid')(x)

# Define the model
model = Model(inputs=[input_statement, input_tweet, input_numerical], outputs=output)

# Compile the model
model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

In [17]:
# Split data into training and test sets
X_train_statement, X_test_statement, X_train_tweet, X_test_tweet, X_train_num, X_test_num, y_train, y_test = train_test_split(
    X_statement_embeddings, X_tweet_embeddings, X_num_scaled, df['majority_target'], test_size=0.3, random_state=42)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=2)
early_stopping = EarlyStopping(monitor='val_loss', patience=3)

# Train the model
history = model.fit([X_train_statement, X_train_tweet, X_train_num], y_train, epochs=10, batch_size=32, validation_split=0.2, callbacks=[early_stopping, reduce_lr])

# Evaluate the model
loss, accuracy = model.evaluate([X_test_statement, X_test_tweet, X_test_num], y_test)
print(f'Test Accuracy: {accuracy}')

# Predict on test data
y_pred = model.predict([X_test_statement, X_test_tweet, X_test_num])

Epoch 1/10


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_22', 'keras_tensor_23', 'keras_tensor_24']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


2349/2349 ━━━━━━━━━━━━━━━━━━━━ 18s 6ms/step - accuracy: 0.8956 - loss: 0.3037 - val_accuracy: 0.9557 - val_loss: 0.1846 - learning_rate: 0.0010
Epoch 2/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 13s 6ms/step - accuracy: 0.9564 - loss: 0.2034 - val_accuracy: 0.9558 - val_loss: 0.1852 - learning_rate: 0.0010
Epoch 3/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 14s 6ms/step - accuracy: 0.9565 - loss: 0.2000 - val_accuracy: 0.9558 - val_loss: 0.1834 - learning_rate: 0.0010
Epoch 4/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 14s 6ms/step - accuracy: 0.9556 - loss: 0.1995 - val_accuracy: 0.9558 - val_loss: 0.1831 - learning_rate: 0.0010
Epoch 5/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 13s 6ms/step - accuracy: 0.9572 - loss: 0.1934 - val_accuracy: 0.9558 - val_loss: 0.1839 - learning_rate: 0.0010
Epoch 6/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 13s 6ms/step - accuracy: 0.9579 - loss: 0.1878 - val_accuracy: 0.9558 - val_loss: 0.1840 - learning_rate: 0.0010
Epoch 7/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 14s 6ms/step - accuracy: 0.9560 - loss:

In [18]:
# Split data into training and test sets
X_train_statement, X_test_statement, X_train_tweet, X_test_tweet, X_train_num, X_test_num, y_train, y_test = train_test_split(
    X_statement_embeddings, X_tweet_embeddings, X_num_scaled, df['majority_target'], test_size=0.3, random_state=42)

reduce_lr = ReduceLROnPlateau(monitor='val_loss', factor=0.2, patience=2)
early_stopping = EarlyStopping(monitor='val_loss', patience=3)

# Train the model
history = model.fit([X_train_statement, X_train_tweet, X_train_num], y_train, epochs=10, batch_size=32, validation_split=0.2, callbacks=[early_stopping, reduce_lr])

# Evaluate the model
loss, accuracy = model.evaluate([X_test_statement, X_test_tweet, X_test_num], y_test)
print(f'Test Accuracy: {accuracy}')

# Predict on test data
y_pred = model.predict([X_test_statement, X_test_tweet, X_test_num])

Epoch 1/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 14s 6ms/step - accuracy: 0.9573 - loss: 0.1840 - val_accuracy: 0.9558 - val_loss: 0.1878 - learning_rate: 2.0000e-04
Epoch 2/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 15s 6ms/step - accuracy: 0.9572 - loss: 0.1798 - val_accuracy: 0.9558 - val_loss: 0.1872 - learning_rate: 2.0000e-04
Epoch 3/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 16s 7ms/step - accuracy: 0.9586 - loss: 0.1757 - val_accuracy: 0.9558 - val_loss: 0.1889 - learning_rate: 2.0000e-04
Epoch 4/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 15s 7ms/step - accuracy: 0.9577 - loss: 0.1782 - val_accuracy: 0.9558 - val_loss: 0.1892 - learning_rate: 2.0000e-04
Epoch 5/10
2349/2349 ━━━━━━━━━━━━━━━━━━━━ 16s 7ms/step - accuracy: 0.9571 - loss: 0.1788 - val_accuracy: 0.9558 - val_loss: 0.1891 - learning_rate: 4.0000e-05
1259/1259 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step - accuracy: 0.9575 - loss: 0.1847
Test Accuracy: 0.9576005935668945
1259/1259 ━━━━━━━━━━━━━━━━━━━━ 3s 2ms/step


In [19]:
y_pred_use = (y_pred > 0.5).astype(int)
# Evaluate the model using accuracy
accuracy = accuracy_score(y_test, y_pred_use)
print(f"RNN Accuracy: {accuracy:.4f}")

print("\nClassification Report:")
print(classification_report(y_test, y_pred_use, target_names=['Fake', 'Real']))

# Confusion Matrix to understand predictions
print("\nConfusion Matrix:")
print(confusion_matrix(y_test, y_pred_use))

RNN Accuracy: 0.9576

Classification Report:
              precision    recall  f1-score   support

        Fake       0.95      0.96      0.96     19442
        Real       0.96      0.96      0.96     20818

    accuracy                           0.96     40260
   macro avg       0.96      0.96      0.96     40260
weighted avg       0.96      0.96      0.96     40260


Confusion Matrix:
[[18618   824]
 [  883 19935]]


Summary of results

basic tokenizer gives 0.9559

with sentence embeddings:

- the one with regularizer gives 0.9571 accuracy
- basic with only dropout is 0.9576
- with BatchNormalization is 0.9576
- with reduce learning rate and early stopping is also 0.9576
- Bi-directional RNN is 0.9576

K-fold cross validation

In [20]:
from sklearn.model_selection import KFold
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score, roc_auc_score, classification_report, confusion_matrix
import numpy as np

# Define number of folds for cross-validation
n_folds = 5
kf = KFold(n_splits=n_folds, shuffle=True, random_state=42)

# Initialize lists to store metrics for each fold
accuracy_list, precision_list, recall_list, f1_list, auc_list = [], [], [], [], []

for train_index, val_index in kf.split(X_statement_embeddings):
    # Split data into training and validation sets for this fold
    X_train_statement, X_val_statement = X_statement_embeddings[train_index], X_statement_embeddings[val_index]
    X_train_tweet, X_val_tweet = X_tweet_embeddings[train_index], X_tweet_embeddings[val_index]
    X_train_num, X_val_num = X_num_scaled[train_index], X_num_scaled[val_index]
    y_train, y_val = df['majority_target'].iloc[train_index], df['majority_target'].iloc[val_index]
    
    # Define inputs and model
    input_statement = Input(shape=(X_statement_embeddings.shape[1],))
    input_tweet = Input(shape=(X_tweet_embeddings.shape[1],))
    input_numerical = Input(shape=(X_num_scaled.shape[1],))
    
    # Reshape to 3D for LSTM (adding a dimension for timesteps)
    reshaped_statement = Reshape((1, X_statement_embeddings.shape[1]))(input_statement)
    reshaped_tweet = Reshape((1, X_tweet_embeddings.shape[1]))(input_tweet)
    
    # LSTM for statement and tweet embeddings
    lstm_statement = LSTM(128, return_sequences=False)(reshaped_statement)
    lstm_tweet = LSTM(128, return_sequences=False)(reshaped_tweet)
    
    # Combine LSTM outputs and numerical features
    combined = Concatenate()([lstm_statement, lstm_tweet, input_numerical])
    
    # Dense layers
    x = Dense(64, activation='relu')(combined)
    x = Dropout(0.5)(x)
    x = Dense(32, activation='relu')(x)
    x = Dropout(0.5)(x)
    output = Dense(1, activation='sigmoid')(x)
    
    # Define and compile model
    model = Model(inputs=[input_statement, input_tweet, input_numerical], outputs=output)
    model.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])
    
    # Train the model
    model.fit([X_train_statement, X_train_tweet, X_train_num], y_train, epochs=10, batch_size=32, verbose=0)
    
    # Predict on validation data
    y_pred_prob = model.predict([X_val_statement, X_val_tweet, X_val_num])
    y_pred = (y_pred_prob > 0.5).astype(int)
    
    # Calculate and store metrics
    accuracy = accuracy_score(y_val, y_pred)
    precision = precision_score(y_val, y_pred)
    recall = recall_score(y_val, y_pred)
    f1 = f1_score(y_val, y_pred)
    auc = roc_auc_score(y_val, y_pred_prob)
    
    accuracy_list.append(accuracy)
    precision_list.append(precision)
    recall_list.append(recall)
    f1_list.append(f1)
    auc_list.append(auc)

# Display the results for each metric
print(f"Cross-Validation Results ({n_folds} folds):")
print(f"Accuracy: {np.mean(accuracy_list):.4f} ± {np.std(accuracy_list):.4f}")
print(f"Precision: {np.mean(precision_list):.4f} ± {np.std(precision_list):.4f}")
print(f"Recall: {np.mean(recall_list):.4f} ± {np.std(recall_list):.4f}")
print(f"F1 Score: {np.mean(f1_list):.4f} ± {np.std(f1_list):.4f}")
print(f"AUC: {np.mean(auc_list):.4f} ± {np.std(auc_list):.4f}")


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_35', 'keras_tensor_36', 'keras_tensor_37']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


839/839 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_48', 'keras_tensor_49', 'keras_tensor_50']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


839/839 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_61', 'keras_tensor_62', 'keras_tensor_63']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


839/839 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_74', 'keras_tensor_75', 'keras_tensor_76']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


839/839 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step


c:\Users\laiye\AppData\Local\Programs\Python\Python311\Lib\site-packages\keras\src\models\functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor_87', 'keras_tensor_88', 'keras_tensor_89']. Received: the structure of inputs=('*', '*', '*')
  warnings.warn(


839/839 ━━━━━━━━━━━━━━━━━━━━ 2s 2ms/step
Cross-Validation Results (5 folds):
Accuracy: 0.9572 ± 0.0006
Precision: 0.9588 ± 0.0021
Recall: 0.9580 ± 0.0013
F1 Score: 0.9584 ± 0.0006
AUC: 0.9585 ± 0.0002


Model Robustness: The model is robust and performs consistently well across different data splits, showing it has generalized well to the dataset.

Performance Ceiling: Since the accuracy, F1 score, and AUC are all very close to 0.96, it’s likely that the model reached a performance ceiling with this dataset and feature set. Additional tuning may not yield significant improvements because the model has effectively captured the key patterns for distinguishing fake from real news.

Balanced Metrics: The high scores in precision and recall, along with a closely matching F1 score, indicate that the model maintains a good balance between identifying both real and fake news accurately, with a strong AUC showing good separation between classes.